File: java/src/main/java/org/example/Utility/CheckSum.java
package org.example.Utility;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.List;

import software.amazon.awssdk.checksums.DefaultChecksumAlgorithm;
import software.amazon.awssdk.checksums.SdkChecksum;
import software.amazon.awssdk.core.SdkField;
import software.amazon.awssdk.core.SdkPojo;
import software.amazon.awssdk.services.s3.model.ChecksumAlgorithm;

public class CheckSum {

	private CheckSum() {
	} // utility class이므로 인스턴스화 방지

	/**
	 * FULL_OBJECT 타입을 지원하는 체크섬 알고리즘 목록 (CRC 계열만 결합 가능)
	 */
	public static final List<ChecksumAlgorithm> FULL_OBJECT_ALGORITHMS = List.of(
			ChecksumAlgorithm.CRC32,
			ChecksumAlgorithm.CRC32_C,
			ChecksumAlgorithm.CRC64_NVME);

	/**
	 * COMPOSITE 타입을 지원하는 체크섬 알고리즘 목록 (CRC64NVME 제외)
	 */
	public static final List<ChecksumAlgorithm> COMPOSITE_ALGORITHMS = List.of(
			ChecksumAlgorithm.CRC32,
			ChecksumAlgorithm.CRC32_C,
			ChecksumAlgorithm.SHA1,
			ChecksumAlgorithm.SHA256,
			ChecksumAlgorithm.MD5,
			ChecksumAlgorithm.SHA512,
			ChecksumAlgorithm.XXHASH64,
			ChecksumAlgorithm.XXHASH3,
			ChecksumAlgorithm.XXHASH128);

	/**
	 * 모든 체크섬 알고리즘 목록 (PutObject, UploadPart 파트 체크섬에서 사용 가능)
	 */
	public static final List<ChecksumAlgorithm> ALL_ALGORITHMS = List.of(
			ChecksumAlgorithm.CRC32,
			ChecksumAlgorithm.CRC32_C,
			ChecksumAlgorithm.CRC64_NVME,
			ChecksumAlgorithm.SHA1,
			ChecksumAlgorithm.SHA256,
			ChecksumAlgorithm.MD5,
			ChecksumAlgorithm.SHA512,
			ChecksumAlgorithm.XXHASH64,
			ChecksumAlgorithm.XXHASH3,
			ChecksumAlgorithm.XXHASH128);

	/**
	 * 문자열에 대한 체크섬 계산
	 */
	public static String calculateChecksum(ChecksumAlgorithm algorithm, String content) {
		return Base64.getEncoder().encodeToString(calculateChecksumBytes(algorithm, content));
	}

	/**
	 * List<byte[]>에 대한 체크섬 계산
	 */
	public static String calculateChecksum(ChecksumAlgorithm algorithm, List<byte[]> contents) {
		return Base64.getEncoder().encodeToString(calculateChecksumBytes(algorithm, contents)) + "-" + contents.size();
	}

	/**
	 * 문자열에 대한 체크섬 bytes 반환
	 */
	public static byte[] calculateChecksumBytes(ChecksumAlgorithm algorithm, String content) {
		SdkChecksum hasher = SdkChecksum.forAlgorithm(DefaultChecksumAlgorithm.fromValue(algorithm.toString()));
		hasher.update(content.getBytes(StandardCharsets.UTF_8));
		return hasher.getChecksumBytes();
	}

	/**
	 * List<byte[]>에 대한 체크섬 bytes 반환
	 */
	public static byte[] calculateChecksumBytes(ChecksumAlgorithm algorithm, List<byte[]> contents) {
		SdkChecksum hasher = SdkChecksum.forAlgorithm(DefaultChecksumAlgorithm.fromValue(algorithm.toString()));
		for (byte[] content : contents) {
			hasher.update(content);
		}
		return hasher.getChecksumBytes();
	}

	/**
	 * List<String>에 대한 체크섬 계산
	 */
	public static String calculateChecksumByBase64(ChecksumAlgorithm algorithm, List<String> contents) {
		return calculateChecksum(algorithm, contents.stream().map(Base64.getDecoder()::decode).toList());
	}

	/**
	 * List<String>에 대한 체크섬 계산
	 */
	public static String combineChecksumByBase64(ChecksumAlgorithm algorithm, long partSize, List<String> contents) {
		byte[] crc1 = Base64.getDecoder().decode(contents.get(0));

		for (var index = 1; index < contents.size(); index++) {
			byte[] crc2 = Base64.getDecoder().decode(contents.get(index));
			crc1 = CrcCombine.combinebytes(crc1, crc2, partSize, algorithm);
		}
		return Base64.getEncoder().encodeToString(crc1);
	}

	/**
	 * 응답 및 모델 객체(PutObjectResponse, UploadPartResponse, CompletedPart,
	 * CopyObjectResult 등)에서 지정한 알고리즘의 체크섬 값을 조회
	 */
	public static String getChecksum(SdkPojo pojo, ChecksumAlgorithm algorithm) {
		return (String) checksumField(pojo, algorithm).getValueOrDefault(pojo);
	}

	/**
	 * 리퀘스트/모델 빌더(CompletedPart.Builder, PutObjectRequest.Builder 등)에
	 * 지정한 알고리즘의 체크섬 값을 설정
	 */
	public static void setChecksum(SdkPojo builder, ChecksumAlgorithm algorithm, String value) {
		checksumField(builder, algorithm).set(builder, value);
	}

	/**
	 * 업로드 리퀘스트 빌더(PutObjectRequest.Builder, UploadPartRequest.Builder 등)에
	 * 체크섬 알고리즘을 지정. 단 SDK가 MD5의 자동 계산을 거부하므로
	 * ("MD5 is not supported. Please use a pre-calculated MD5 value")
	 * MD5는 본문으로 사전 계산한 체크섬 값을 대신 지정한다.
	 */
	public static void applyChecksum(SdkPojo builder, ChecksumAlgorithm algorithm, String content) {
		if (algorithm == ChecksumAlgorithm.MD5) {
			setChecksum(builder, algorithm, calculateChecksum(algorithm, content));
		} else {
			findField(builder, "ChecksumAlgorithm").set(builder, algorithm.toString());
		}
	}

	private static SdkField<?> checksumField(SdkPojo pojo, ChecksumAlgorithm algorithm) {
		return findField(pojo, "Checksum" + algorithm);
	}

	private static SdkField<?> findField(SdkPojo pojo, String memberName) {
		for (SdkField<?> field : pojo.sdkFields()) {
			if (field.memberName().equals(memberName)) {
				return field;
			}
		}
		throw new IllegalArgumentException("Field not found: " + memberName);
	}

	/**
	 * List<byte[]>에 대한 CRC32 체크섬 계산
	 */
	public static String crc32(List<byte[]> contents) {
		return calculateChecksum(ChecksumAlgorithm.CRC32, contents);
	}

	/**
	 * List<byte[]>에 대한 CRC32C 체크섬 계산
	 */
	public static String crc32c(List<byte[]> contents) {
		return calculateChecksum(ChecksumAlgorithm.CRC32_C, contents);
	}

	/**
	 * List<byte[]>에 대한 CRC64 체크섬 계산
	 */
	public static String crc64Nvme(List<byte[]> contents) {
		return calculateChecksum(ChecksumAlgorithm.CRC64_NVME, contents);
	}

	/**
	 * List<byte[]>에 대한 SHA1 체크섬 계산
	 */
	public static String sha1(List<byte[]> contents) {
		return calculateChecksum(ChecksumAlgorithm.SHA1, contents);
	}

	/**
	 * List<byte[]>에 대한 SHA256 체크섬 계산
	 */
	public static String sha256(List<byte[]> contents) {
		return calculateChecksum(ChecksumAlgorithm.SHA256, contents);
	}
}

File: dotnet/Utils/CheckSum.cs
using System;
using System.Collections.Generic;
using System.IO.Hashing;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using Amazon.S3;
using Amazon.S3.Model;

namespace s3tests.Utils
{
	/// <summary>
	/// S3 checksum helpers. Ported from java CheckSum.java.
	/// .NET AWSSDK.S3 supports CRC32, CRC32C, CRC64NVME, SHA1, SHA256.
	/// </summary>
	public static class CheckSum
	{
		public static readonly IReadOnlyList<ChecksumAlgorithm> FullObjectAlgorithms =
		[
			ChecksumAlgorithm.CRC32,
			ChecksumAlgorithm.CRC32C,
			ChecksumAlgorithm.CRC64NVME,
		];

		public static readonly IReadOnlyList<ChecksumAlgorithm> CompositeAlgorithms =
		[
			ChecksumAlgorithm.CRC32,
			ChecksumAlgorithm.CRC32C,
			ChecksumAlgorithm.SHA1,
			ChecksumAlgorithm.SHA256,
		];

		public static readonly IReadOnlyList<ChecksumAlgorithm> AllAlgorithms =
		[
			ChecksumAlgorithm.CRC32,
			ChecksumAlgorithm.CRC32C,
			ChecksumAlgorithm.CRC64NVME,
			ChecksumAlgorithm.SHA1,
			ChecksumAlgorithm.SHA256,
		];

		public static string CalculateChecksum(ChecksumAlgorithm algorithm, string content)
			=> Convert.ToBase64String(CalculateChecksumBytes(algorithm, Encoding.UTF8.GetBytes(content)));

		public static string CalculateChecksum(ChecksumAlgorithm algorithm, IList<byte[]> contents)
			=> Convert.ToBase64String(CalculateChecksumBytes(algorithm, contents)) + "-" + contents.Count;

		public static byte[] CalculateChecksumBytes(ChecksumAlgorithm algorithm, byte[] content)
		{
			if (algorithm == ChecksumAlgorithm.CRC32)
			{
				var crc = new Crc32();
				crc.Append(content);
				return ToBigEndian(crc.GetCurrentHash());
			}
			if (algorithm == ChecksumAlgorithm.CRC32C)
				return ComputeCrc32C(content);
			if (algorithm == ChecksumAlgorithm.CRC64NVME)
				return ComputeCrc64Nvme(content);
			if (algorithm == ChecksumAlgorithm.SHA1)
				return SHA1.HashData(content);
			if (algorithm == ChecksumAlgorithm.SHA256)
				return SHA256.HashData(content);

			throw new ArgumentException($"Unsupported checksum algorithm: {algorithm}");
		}

		public static byte[] CalculateChecksumBytes(ChecksumAlgorithm algorithm, string content)
			=> CalculateChecksumBytes(algorithm, Encoding.UTF8.GetBytes(content));

		public static byte[] CalculateChecksumBytes(ChecksumAlgorithm algorithm, IList<byte[]> contents)
		{
			var hasher = CreateIncremental(algorithm);
			foreach (var part in contents)
				hasher.Append(part);
			return hasher.GetHashAndReset();
		}

		public static string CalculateChecksumByBase64(ChecksumAlgorithm algorithm, IList<string> contents)
		{
			var decoded = contents.Select(Convert.FromBase64String).ToList();
			return CalculateChecksum(algorithm, decoded);
		}

		public static string CombineChecksumByBase64(ChecksumAlgorithm algorithm, long partSize, IList<string> contents)
		{
			byte[] crc1 = Convert.FromBase64String(contents[0]);
			for (int i = 1; i < contents.Count; i++)
			{
				byte[] crc2 = Convert.FromBase64String(contents[i]);
				crc1 = CrcCombine.CombineBytes(crc1, crc2, partSize, algorithm);
			}
			return Convert.ToBase64String(crc1);
		}

		public static string GetChecksum(GetObjectAttributesResponse response, ChecksumAlgorithm algorithm)
			=> GetChecksumValue(response?.Checksum?.ChecksumCRC32, response?.Checksum?.ChecksumCRC32C,
				response?.Checksum?.ChecksumCRC64NVME, response?.Checksum?.ChecksumSHA1,
				response?.Checksum?.ChecksumSHA256, algorithm);

		public static string GetChecksum(PutObjectResponse response, ChecksumAlgorithm algorithm)
			=> GetChecksumValue(response.ChecksumCRC32, response.ChecksumCRC32C, response.ChecksumCRC64NVME,
				response.ChecksumSHA1, response.ChecksumSHA256, algorithm);

		public static string GetChecksum(CopyObjectResponse response, ChecksumAlgorithm algorithm)
			=> GetChecksumValue(response.ChecksumCRC32, response.ChecksumCRC32C, response.ChecksumCRC64NVME,
				response.ChecksumSHA1, response.ChecksumSHA256, algorithm);

		public static string GetChecksum(UploadPartResponse response, ChecksumAlgorithm algorithm)
			=> GetChecksumValue(response.ChecksumCRC32, response.ChecksumCRC32C, response.ChecksumCRC64NVME,
				response.ChecksumSHA1, response.ChecksumSHA256, algorithm);

		public static string GetChecksum(CompleteMultipartUploadResponse response, ChecksumAlgorithm algorithm)
			=> GetChecksumValue(response.ChecksumCRC32, response.ChecksumCRC32C, response.ChecksumCRC64NVME,
				response.ChecksumSHA1, response.ChecksumSHA256, algorithm);

		public static string GetChecksum(PartETag part, ChecksumAlgorithm algorithm)
			=> GetChecksumValue(part.ChecksumCRC32, part.ChecksumCRC32C, part.ChecksumCRC64NVME,
				part.ChecksumSHA1, part.ChecksumSHA256, algorithm);

		public static void SetChecksum(PutObjectRequest request, ChecksumAlgorithm algorithm, string value)
		{
			if (algorithm == ChecksumAlgorithm.CRC32) request.ChecksumCRC32 = value;
			else if (algorithm == ChecksumAlgorithm.CRC32C) request.ChecksumCRC32C = value;
			else if (algorithm == ChecksumAlgorithm.CRC64NVME) request.ChecksumCRC64NVME = value;
			else if (algorithm == ChecksumAlgorithm.SHA1) request.ChecksumSHA1 = value;
			else if (algorithm == ChecksumAlgorithm.SHA256) request.ChecksumSHA256 = value;
			else throw new ArgumentException($"Unsupported: {algorithm}");
		}

		public static void SetChecksum(UploadPartRequest request, ChecksumAlgorithm algorithm, string value)
		{
			if (algorithm == ChecksumAlgorithm.CRC32) request.ChecksumCRC32 = value;
			else if (algorithm == ChecksumAlgorithm.CRC32C) request.ChecksumCRC32C = value;
			else if (algorithm == ChecksumAlgorithm.CRC64NVME) request.ChecksumCRC64NVME = value;
			else if (algorithm == ChecksumAlgorithm.SHA1) request.ChecksumSHA1 = value;
			else if (algorithm == ChecksumAlgorithm.SHA256) request.ChecksumSHA256 = value;
			else throw new ArgumentException($"Unsupported: {algorithm}");
		}

		public static void SetChecksum(PartETag part, ChecksumAlgorithm algorithm, string value)
		{
			if (algorithm == ChecksumAlgorithm.CRC32) part.ChecksumCRC32 = value;
			else if (algorithm == ChecksumAlgorithm.CRC32C) part.ChecksumCRC32C = value;
			else if (algorithm == ChecksumAlgorithm.CRC64NVME) part.ChecksumCRC64NVME = value;
			else if (algorithm == ChecksumAlgorithm.SHA1) part.ChecksumSHA1 = value;
			else if (algorithm == ChecksumAlgorithm.SHA256) part.ChecksumSHA256 = value;
			else throw new ArgumentException($"Unsupported: {algorithm}");
		}

		public static void ApplyChecksum(PutObjectRequest request, ChecksumAlgorithm algorithm)
			=> request.ChecksumAlgorithm = algorithm;

		public static void ApplyChecksum(UploadPartRequest request, ChecksumAlgorithm algorithm)
			=> request.ChecksumAlgorithm = algorithm;

		public static void ApplyChecksum(InitiateMultipartUploadRequest request, ChecksumAlgorithm algorithm, ChecksumType checksumType = null)
		{
			request.ChecksumAlgorithm = algorithm;
			if (checksumType != null) request.ChecksumType = checksumType;
		}

		private static string GetChecksumValue(string crc32, string crc32c, string crc64, string sha1, string sha256,
			ChecksumAlgorithm algorithm)
		{
			if (algorithm == ChecksumAlgorithm.CRC32) return crc32;
			if (algorithm == ChecksumAlgorithm.CRC32C) return crc32c;
			if (algorithm == ChecksumAlgorithm.CRC64NVME) return crc64;
			if (algorithm == ChecksumAlgorithm.SHA1) return sha1;
			if (algorithm == ChecksumAlgorithm.SHA256) return sha256;
			throw new ArgumentException($"Unsupported: {algorithm}");
		}

		/// <summary>
		/// System.IO.Hashing의 CRC 계열은 해시를 리틀엔디언으로 돌려주지만
		/// S3 체크섬 헤더는 빅엔디언(네트워크 바이트 순서)을 기대한다.
		/// </summary>
		private static byte[] ToBigEndian(byte[] hash)
		{
			var bytes = (byte[])hash.Clone();
			Array.Reverse(bytes);
			return bytes;
		}

		private static byte[] ComputeCrc32C(byte[] content)
		{
			// System.IO.Hashing 9.x has Crc32 (IEEE). Use reflection for Crc32C if present.
			var crc32cType = Type.GetType("System.IO.Hashing.Crc32C, System.IO.Hashing");
			if (crc32cType != null)
			{
				dynamic instance = Activator.CreateInstance(crc32cType);
				instance.Append(content);
				return ToBigEndian((byte[])instance.GetCurrentHash());
			}
			// Fallback: Force.Crc32 package-style Castagnoli via manual if needed
			return ComputeCrc32CManual(content);
		}

		private static byte[] ComputeCrc32CManual(byte[] data)
		{
			uint crc = 0xFFFFFFFFu;
			foreach (byte b in data)
			{
				crc ^= b;
				for (int i = 0; i < 8; i++)
					crc = (crc & 1) != 0 ? (0x82F63B78u ^ (crc >> 1)) : (crc >> 1);
			}
			crc ^= 0xFFFFFFFFu;
			var bytes = BitConverter.GetBytes(crc);
			if (BitConverter.IsLittleEndian) Array.Reverse(bytes);
			return bytes;
		}

		// CRC64-NVME polynomial 0x9A6C9329AC4BC9B5 (reflected), init/xor 0xFFFFFFFFFFFFFFFF
		private static byte[] ComputeCrc64Nvme(byte[] data)
		{
			ulong crc = 0xFFFFFFFFFFFFFFFFul;
			foreach (byte b in data)
			{
				crc ^= b;
				for (int i = 0; i < 8; i++)
					crc = (crc & 1) != 0 ? (0x9A6C9329AC4BC9B5ul ^ (crc >> 1)) : (crc >> 1);
			}
			crc ^= 0xFFFFFFFFFFFFFFFFul;
			var bytes = BitConverter.GetBytes(crc);
			if (BitConverter.IsLittleEndian) Array.Reverse(bytes);
			return bytes;
		}

		private interface IIncrementalHash
		{
			void Append(byte[] data);
			byte[] GetHashAndReset();
		}

		private static IIncrementalHash CreateIncremental(ChecksumAlgorithm algorithm)
		{
			if (algorithm == ChecksumAlgorithm.SHA1) return new IncrementalSha1();
			if (algorithm == ChecksumAlgorithm.SHA256) return new IncrementalSha256();
			if (algorithm == ChecksumAlgorithm.CRC32) return new IncrementalCrc32();
			if (algorithm == ChecksumAlgorithm.CRC32C) return new IncrementalCrc32C();
			if (algorithm == ChecksumAlgorithm.CRC64NVME) return new IncrementalCrc64Nvme();
			throw new ArgumentException($"Unsupported: {algorithm}");
		}

		private sealed class IncrementalSha1 : IIncrementalHash
		{
			private readonly IncrementalHash _hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA1);
			public void Append(byte[] data) => _hash.AppendData(data);
			public byte[] GetHashAndReset() => _hash.GetHashAndReset();
		}

		private sealed class IncrementalSha256 : IIncrementalHash
		{
			private readonly IncrementalHash _hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
			public void Append(byte[] data) => _hash.AppendData(data);
			public byte[] GetHashAndReset() => _hash.GetHashAndReset();
		}

		private sealed class IncrementalCrc32 : IIncrementalHash
		{
			private readonly List<byte> _buf = [];
			public void Append(byte[] data) => _buf.AddRange(data);
			public byte[] GetHashAndReset()
			{
				var crc = new Crc32();
				crc.Append(_buf.ToArray());
				_buf.Clear();
				return ToBigEndian(crc.GetCurrentHash());
			}
		}

		private sealed class IncrementalCrc32C : IIncrementalHash
		{
			private readonly List<byte> _buf = [];
			public void Append(byte[] data) => _buf.AddRange(data);
			public byte[] GetHashAndReset()
			{
				var result = ComputeCrc32C(_buf.ToArray());
				_buf.Clear();
				return result;
			}
		}

		private sealed class IncrementalCrc64Nvme : IIncrementalHash
		{
			private readonly List<byte> _buf = [];
			public void Append(byte[] data) => _buf.AddRange(data);
			public byte[] GetHashAndReset()
			{
				var result = ComputeCrc64Nvme(_buf.ToArray());
				_buf.Clear();
				return result;
			}
		}
	}
}
